nvflare/fed_job.py

```diff
@@ -0,0 +1,274 @@
+            id = f"{id}_{cnt}"
+            self._used_ids.append(id)
+        return id
+
+
+class FedJob:
+    def __init__(self, name="fed_job", min_clients=1, mandatory_clients=None, key_metric="accuracy") -> None:
+        """FedJob allows users to generate job configurations in a Pythonic way.
+        The `to()` routine allows users to send different components to either the server or clients.
+
+        Args:
+            job_name: the name of the NVFlare job
+            min_clients: the minimum number of clients for the job
+            mandatory_clients: mandatory clients to run the job (optional)
+            key_metric: Metric used to determine if the model is globally best.
+                if metrics are a `dict`, `key_metric` can select the metric used for global model selection.
+                Defaults to "accuracy".
+        """
+        self.job_name = name
+        self.key_metric = key_metric
+        self.clients = []
+        self.job: FedJobConfig = FedJobConfig(
+            job_name=self.job_name, min_clients=min_clients, mandatory_clients=mandatory_clients
+        )
+        self._deploy_map = {}
+        self._deployed = False
+        self._gpus = {}
+
+    def to(
+        self, obj: Any, target: str, tasks: List[str] = None, gpu: int = None, filter_type: FilterType = None, id=None
+    ):
+        """assign an `obj` to a target (server or clients).
+        The obj will be given a default `id` if non is provided based on its type.
+
+        Returns:
+
+        """
+        if isinstance(obj, Controller):
+            if target not in self._deploy_map:
+                self._deploy_map[target] = ControllerApp(key_metric=self.key_metric)
+            self._deploy_map[target].add_controller(obj, id)
+        elif isinstance(obj, Executor):
+            if target not in self._deploy_map:
+                if isinstance(obj, ScriptExecutor):
+                    external_scripts = [obj._task_script_path]
+                else:
+                    external_scripts = None
+                self._deploy_map[target] = ExecutorApp(external_scripts=external_scripts)
+                self.clients.append(target)
+                if gpu is not None:
+                    if target not in self._gpus:  # GPU can only be selected once per client.
+                        self._gpus[target] = str(gpu)
+                    else:
+                        print(f"{target} already set to use GPU {self._gpus[target]}. Ignoring gpu={gpu}.")
+            self._deploy_map[target].add_executor(obj, tasks=tasks)
+        else:  # handle objects that are not Controller or Executor type
+            if target not in self._deploy_map:
+                raise ValueError(
+                    f"{target} doesn't have a `Controller` or `Executor`. Deploy one first before adding components!"
+                )
+
+            if isinstance(obj, nn.Module):  # if model, set a persistor
+                self._deploy_map[target].set_persistor(obj)
```

Comment:
This particular model handling does not belong to fed_job.py.

Reply:
I refactored this class to use optional imports. Please take a look.